smarty/sampler_smirky.py

```diff
@@ -243,9 +314,37 @@ def __init__(self, molecules, typetag, elementList, ORdecorators,
             for label_set in self.ref_labels:
                 for (atom_indices, pid, smirks) in label_set:
                     self.reference_type_counts[pid] += 1
-            if self.verbose: self.show_type_statistics(typelist, typecounts, molecule_typecounts, self.type_matches)
+            self.write_type_statistics(typelist, typecounts, molecule_typecounts, self.type_matches)
         return
 
+    def _makeEnvironments(self, typetag, smirksList):
+        """
+        Given a typetag and list of SMIRKS strings
+        returns a list of chemical environment objects
+        of the correct type
+        """
+        if typetag.lower() == 'vdw':
+            chemEnv = AtomChemicalEnvironment
+        elif typetag.lower() == 'bond':
+            chemEnv = BondChemicalEnvironment
+        elif typetag.lower() == 'angle':
+            chemEnv = AngleChemicalEnvironment
+        elif typetag.lower() == 'torsion':
+            chemEnv = TorsionChemicalEnvironment
+        elif typetag.lower() == 'improper':
+            chemEnv = ImproperChemicalEnvironment
+        else:
+            return None
+
+        envList = list()
+        if smirksList is None:
+            return [chemEnv(None, 0, self.replacements)]
+
+        for smirks, typename in smirksList:
+            envList.append(chemEnv(smirks, typename, self.replacements))
+
+        return envList
+
     def get_force_type(self, typetag):
         """
         Uses typetag to get the Force type key word used to read the SMIRFF file
```

Comment:
You have a comment in the code on line 366 asking what the force word is for impropers. Impropers are a `PeriodicTorsionForce` (they still exert a force via that mechanism) so they don't have their own keyword; they're distinguished only by what atoms the force is applied between.